src/main/java/uk/gov/register/store/postgres/PostgresDataAccessLayer.java

```diff
@@ -14,38 +17,79 @@
 public class PostgresDataAccessLayer extends PostgresReadDataAccessLayer implements DataAccessLayer {
     private final List<Entry> stagedEntries;
     private final Map<HashValue, Item> stagedItems;
-    private final HashMap<String, Integer> stagedCurrentKeys;
-    private final Set<String> entriesWithoutItems;
+    private final HashSet<String> stagedEntryKeys;
+    private final Map<String, Map<String, List<StartIndex>>> stagedStartIndexes;
+    private final Map<String, List<EndIndex>> stagedEndIndexes;
 
     private final EntryDAO entryDAO;
     private final EntryItemDAO entryItemDAO;
     private final ItemDAO itemDAO;
     private final IndexDAO indexDAO;
+    private final IndexDriver indexDriver;
+    private final Map<EntryType, Collection<IndexFunction>> indexFunctionsByEntryType;
 
     public PostgresDataAccessLayer(
             EntryQueryDAO entryQueryDAO, IndexDAO indexDAO, IndexQueryDAO indexQueryDAO, EntryDAO entryDAO,
             EntryItemDAO entryItemDAO, ItemQueryDAO itemQueryDAO,
-            ItemDAO itemDAO, String schema) {
+            ItemDAO itemDAO, String schema, IndexDriver indexDriver, Map<EntryType, Collection<IndexFunction>> indexFunctionsByEntryType) {
         super(entryQueryDAO, indexQueryDAO, itemQueryDAO, schema);
         this.entryDAO = entryDAO;
         this.entryItemDAO = entryItemDAO;
         this.itemDAO = itemDAO;
         this.indexDAO = indexDAO;
+        
+        this.indexDriver = indexDriver;
+        this.indexFunctionsByEntryType = indexFunctionsByEntryType;
 
         stagedEntries = new ArrayList<>();
         stagedItems = new HashMap<>();
-        stagedCurrentKeys = new HashMap<>();
-        entriesWithoutItems = new HashSet<>();
+        stagedEntryKeys = new HashSet<>();
+        stagedStartIndexes = new HashMap<>();
+        stagedEndIndexes = new HashMap<>();
     }
 
     @Override
     public void appendEntry(Entry entry) {
         stagedEntries.add(entry);
+        stagedEntryKeys.add(entry.getKey());
 
         if (entry.getEntryType().equals(EntryType.system)) {
             checkpoint();
         }
     }
+    
+    @Override
+    public int getCurrentIndexEntryNumber(String indexName) {
+        return indexQueryDAO.getCurrentIndexEntryNumber(indexName, schema);
+    }
+
+    @Override
+    public IndexEntryNumberItemCountPair getStartIndexEntryNumberAndExistingItemCount(String indexName, String key, String sha256hex){
```

Comment:
This method doesn't account for the fact that the index table could have some rows in it that need to be included in this query. I think this might be a problem?